server/src/main/java/org/elasticsearch/action/get/TransportShardMultiGetAction.java

```diff
@@ -173,57 +182,115 @@ private void handleMultiGetOnUnpromotableShard(
-                        l.onResponse(r.multiGetShardResponse());
-                    } else {
-                        logger.debug(
-                            "no result for some ids from the promotable shard (segment generation to wait for: {})",
-                            r.segmentGeneration()
-                        );
-                        if (r.segmentGeneration() == -1) {
-                            // Nothing to wait for (no previous unsafe generation), just handle the rest locally.
-                            ActionRunnable.supply(l, () -> handleLocalGets(request, r.multiGetShardResponse(), shardId)).run();
-                        } else {
-                            assert r.segmentGeneration() > -1L;
-                            assert r.primaryTerm() > Engine.UNKNOWN_PRIMARY_TERM;
-                            indexShard.waitForPrimaryTermAndGeneration(
-                                r.primaryTerm(),
-                                r.segmentGeneration(),
-                                listener.delegateFailureAndWrap(
-                                    (ll, aLong) -> getExecutor(request, shardId).execute(
-                                        ActionRunnable.supply(ll, () -> handleLocalGets(request, r.multiGetShardResponse(), shardId))
-                                    )
-                                )
-                            );
-                        }
-                    }
-                }), TransportShardMultiGetFomTranslogAction.Response::new, getExecutor(request, shardId))
+            return;
+        }
+        if (request.realtime()) {
+            final var state = clusterService.state();
+            final var observer = new ClusterStateObserver(
+                state,
+                clusterService,
+                TimeValue.timeValueSeconds(60),
+                logger,
+                threadPool.getThreadContext()
             );
+            shardMultiGetFromTranslog(request, indexShard, state, observer, listener);
         } else {
             // A non-real-time mget with no explicit refresh requested.
             super.asyncShardOperation(request, shardId, listener);
         }
     }
 
+    private void shardMultiGetFromTranslog(
+        MultiGetShardRequest request,
+        IndexShard indexShard,
+        ClusterState state,
+        ClusterStateObserver observer,
+        ActionListener<MultiGetShardResponse> listener
+    ) {
+        tryShardMultiGetFromTranslog(request, indexShard, state, listener.delegateResponse((l, e) -> {
+            final var cause = ExceptionsHelper.unwrapCause(e);
+            logger.debug("mget_from_translog[shard] failed", cause);
+            if (cause instanceof ShardNotFoundException
+                || cause instanceof IndexNotFoundException
+                || cause instanceof NoShardAvailableActionException
+                || cause instanceof UnavailableShardsException) {
```

Comment:
Based on this comment https://github.com/elastic/elasticsearch/pull/104579#discussion_r1472954697, I assume we should remove these two exceptions here as well and also make `getCurrentNodeOfPrimary` use listener?

Reply:
ja!